types/wlr_input_method_v2.c

```diff
@@ -108,10 +116,126 @@ static void im_delete_surrounding_text(struct wl_client *client,
 }
 
+bool wlr_surface_is_input_popup_surface_v2(struct wlr_surface *surface) {
+	return surface->role == &input_popup_surface_v2_role;
+}
+
+struct wlr_input_popup_surface_v2 *wlr_input_popup_surface_v2_from_wlr_surface(
+		struct wlr_surface *surface) {
+	assert(wlr_surface_is_input_popup_surface_v2(surface));
+	return (struct wlr_input_popup_surface_v2 *)surface->role_data;
+}
+
+void wlr_input_popup_surface_v2_send_text_input_rectangle(
+		struct wlr_input_popup_surface_v2 *popup_surface,
+		struct wlr_box *sbox) {
+	zwp_input_popup_surface_v2_send_text_input_rectangle(
+		popup_surface->resource, sbox->x, sbox->y, sbox->width, sbox->height);
+}
+
+static void popup_surface_set_mapped(
+		struct wlr_input_popup_surface_v2 *popup_surface, bool mapped) {
+	if (mapped && !popup_surface->mapped) {
+		popup_surface->mapped = true;
+		wlr_signal_emit_safe(&popup_surface->events.map, popup_surface);
+	} else if (!mapped && popup_surface->mapped) {
+		wlr_signal_emit_safe(&popup_surface->events.unmap, popup_surface);
+		popup_surface->mapped = false;
+	}
+}
+
+static void popup_surface_surface_role_precommit(struct wlr_surface *surface) {
+	struct wlr_input_popup_surface_v2 *popup_surface = surface->role_data;
+	if (popup_surface == NULL) {
+		return;
+	}
+	if (surface->pending.committed & WLR_SURFACE_STATE_BUFFER &&
+			surface->pending.buffer_resource == NULL) {
+		// This is a NULL commit
+		popup_surface_set_mapped(popup_surface, false);
+	}
+}
+
+static void popup_surface_surface_role_commit(struct wlr_surface *surface) {
+	struct wlr_input_popup_surface_v2 *popup_surface = surface->role_data;
+	if (popup_surface == NULL) {
+		return;
+	}
+	popup_surface_set_mapped(popup_surface, wlr_surface_has_buffer(surface)
+		&& popup_surface->input_method->client_active);
+}
+
+const struct wlr_surface_role input_popup_surface_v2_role = {
+	.name = "input_popup",
+	.precommit = popup_surface_surface_role_precommit,
+	.commit = popup_surface_surface_role_commit,
+};
+
+static void popup_surface_destroy(
+		struct wlr_input_popup_surface_v2 *popup_surface) {
+	popup_surface_set_mapped(popup_surface, false);
+	wlr_signal_emit_safe(&popup_surface->events.destroy, popup_surface);
+	wl_list_remove(&popup_surface->surface_destroy.link);
+	wl_list_remove(&popup_surface->link);
+	free(popup_surface);
+}
+
+static void popup_surface_handle_surface_destroy(struct wl_listener *listener,
+		void *data) {
+	struct wlr_input_popup_surface_v2 *popup_surface =
+		wl_container_of(listener, popup_surface, surface_destroy);
+	popup_surface_destroy(popup_surface);
+}
+
 static void im_get_input_popup_surface(struct wl_client *client,
 		struct wl_resource *resource, uint32_t id,
 		struct wl_resource *surface) {
-	wlr_log(WLR_INFO, "Stub: zwp_input_method_v2::get_input_popup_surface");
+	struct wlr_input_method_v2 *input_method =
+		input_method_from_resource(resource);
+	if (!input_method) {
+		return;
+	}
+
+	struct wl_resource *popup_resource = wl_resource_create(
```

Comment:
This creates the resource without setting the implementation, however the interface has one request (destroy). We should call `wl_resource_set_implementation` with a handler for this request (which would just call `wl_resource_destroy`).

We also need to make sure the `struct wlr_input_popup_surface_v2` is destroyed when the `wl_resource` is (can happen with the destroy handler described above, but also if the client crashes and the Wayland connection is closed).